### ISSUE TITLE
where is the val set

### ISSUE THREAD

**Issue by ahustr** [naver-ai/pcme#9]
During training it seems that only the training set and the test set are included, but not the validation set When datasetname is cub

**Comment by ahustr** [comment]
Does dataloaders['val_in'] correspond to a test set?

**Comment by SanghyukChun** [comment]
Hello,
`val_in` is the validation samples with the same classes as the train set. `val` is the correct evaluation split.

https://github.com/naver-ai/pcme/blob/7224f04e4f780545176e3bc5040ad9b9e78a5049/datasets/_dataloader.py#L144-L149

You can check the `val` split is used for the evaluation code.
https://github.com/naver-ai/pcme/blob/7224f04e4f780545176e3bc5040ad9b9e78a5049/evaluate_cub.py#L36-L58

Please read the document for more details.
https://github.com/naver-ai/pcme/tree/main#cub-caption-1

**Comment by SanghyukChun** [comment]
For clarification, don't use `cub` for hyperparameter search. My code includes `cub_trainval1` `cub_trainval2` and `cub_trainval3` for the validation.
Check details here:

https://github.com/naver-ai/pcme#hyperparameter-search

**Comment by ahustr** [comment]
I'm sorry but I still don't particularly get it, I observe that MSCOCO sets the train val and test sets separately but for CUB it seems to have only train and val, does this mean that the final result is taken directly from the optimal val of the training process without the process of now validating the val set and then testing it in the test set
![image](https://github.com/naver-ai/pcme/assets/55874332/2dbf9888-5939-4af5-9857-fb15b55ccc2c)
![image](https://github.com/naver-ai/pcme/assets/55874332/62829556-24d7-41b8-9910-a0c6b77d2121)

**Comment by SanghyukChun** [comment]
Okay, so basically, the CUB Caption dataset used in my paper is totally different from MS-COCO Caption. Throughout a series of works [1,2,3,4], I keep arguing that the false negatives (FNs) in the dataset make the evaluation wrong. In other words, although the retrieved item is plausible to human, as the dataset only contains one positive image-caption relationship. Also, at the same time, I argue that recall scores often mislead the performance of the model; we need precision, not recall for better retrieval.

CUB Caption is my primitive approach to tackle these problems. Instead of using the given image-text pairs in the dataset, I defined all images and texts in the same class are positive relathions (e.g., all images and captions for CUB class 1 -- Black_footed_Albatross). It is because the CUB captions describe the appearance of the species in terms of color, shape, pattern, and etc.

Therefore, CUB Caption has very plenty of positives compared to COCO Caption. However, there is a huge problem. As the size of the dataset is too small, the numbers are easily hacked by the hyperparameter choice. It is already observed and well-discussed in zero-shot classification [5]. Therefore, following Xian et al [5], I split the dataset into 150:50 split, where 150 classes are used for training and hyperparameter search, and the remaining 50 classes are used for the test.

(Another clarification: CUB has 200 classes. The 150:50 split does mean that the classes are split, i.e., the classes in the training and test splits are not duplicated)

Now, back to the code again. In the code, you can see that there are `train`, `val_in`, and `val` splits. `val` split is the `50` number of `test` split. `train` and `val_in` are automatically separated samples from `150` classes. The name `val_in` means that it is `validation samples in the known class`. `val_in` is usually used for early stopping.
https://github.com/naver-ai/pcme/blob/7224f04e4f780545176e3bc5040ad9b9e78a5049/datasets/_dataloader.py#L166-L189

However, for the hyperparameter tuning, we sometimes need to verify the generalizability of the model to the unseen classes. To verify this, the 150 seen classes are split into 50:50:50 classes. Each `cub_trainvalX` where `X` is either `{1, 2, 3}`, denotes that each 50 classes are selected as the validation split. Therefore, you can directly tune your hyperparameters on `cub_trainvalX`. If your algorithm is sufficiently robust to the hyperparameter choice, we can believe that it will be generalized to the `cub` setting, i.e., training on 150 classes and testing 50 classes.

In my opinion, the CUB Caption is a good dataset. I considered a lot when I made this dataset. However, if you are not affordable to implement your code on this specific types of annotations, I'd like to suggest you to evaluate your COCO-trained model on ECCV Caption mAP@R [2]. You can check the details in the official repository (https://github.com/naver-ai/eccv-caption)

- [1] Probabilistic Embeddings for Cross-Modal Retrieval. Sanghyuk Chun, Seong Joon Oh, Rafael Sampaio de Rezende, Yannis Kalantidis, Diane Larlus, in CVPR 2021
- [2] ECCV Caption: Correcting False Negatives by Collecting Machine-and-Human-verified Image-Caption Associations for MS-COCO. Sanghyuk Chun, Wonjae Kim, Song Park, Minsuk Chang, Seong Joon Oh, in ECCV 2022
- [3] Improved Probabilistic Image-Text Representations. Sanghyuk Chun, preprint 2023
- [4] LinCIR: Language-only Training of Zero-shot Composed Image Retrieval, Geonmo Gu*, Sanghyuk Chun*, Wonjae Kim, Yoohoon Kang, Sangdoo Yun, preprint 2023
- [5] Zero-Shot Learning -- The Good, the Bad and the Ugly, Yongqin Xian, Bernt Schiele, Zeynep Akata, in CVPR 2017

**Comment by ahustr** [comment]
Thank you very much for your patience, it gives me a sense of the solidity and excellence of your work. I am sure this work will have a profound impact on the field of image-text retrieval. After understanding your answer, I have one last point of confusion, as depicted in the code below, it seems that while saving the models, we save the best model on the test set('val'). This is different from my previous experience where the best model on the validation set was saved and later tested on the test set. The test set will be used only once, i.e. for testing generalization accuracy and not for saving the model.
![image](https://github.com/naver-ai/pcme/assets/55874332/0ee58c22-d68f-415b-9281-4e6d22d5031e)
![image](https://github.com/naver-ai/pcme/assets/55874332/27666309-8995-449b-946b-1813764910a1)

**Comment by SanghyukChun** [comment]
Ah ha, very great question.
My implementation is actually based on `last` model selection, not `best` model selection. Therefore, your concern does not actually happen (because we always choose the `last` model). The `best` model reporting is the legacy of the reference codes (e.g., VSE++, PVSE, ...). It is because I used AdamP + cosine learning rate scheduler, where the last model is (nealy-almost) always the best model.

Maybe it is a out-of-context, but you can check that using the last model selection + AdamP + SWA is practically the best model selection criterion without a validation split (for more details, check my latest work: [3] https://github.com/naver-ai/pcmepp, or the model selection discussion for SWA + last in domain generalization [6]).

- [3] Improved Probabilistic Image-Text Representations. Sanghyuk Chun, preprint 2023
- [6] SWAD: Domain Generalization by Seeking Flat Minima, Junbum Cha, Sanghyuk Chun, Kyungjae Lee, Han-Cheol Cho, Seunghyun Park, Yunsung Lee, Sungrae Park, in NeurIPS 2021

If you want to implement the validation best model selection for the CUB Caption dataset, please use `val_in` for the model selection, not `val` split. I think that the current code does not support `val_in` model selection (forgive me if I'm wrong, it has been four years since I implemented this code)

**Comment by ahustr** [comment]
Thank you very much, I have understood it completely!
